File: events/ccppbrasil_13o_encontro/code/sample-1/rep.c
//Sample provided by Fabio Galuppo  
//July 2017

//compile:
//clang -I../include -L/usr/local/Cellar/zeromq/4.2.2/lib -lzmq rep.c -o ./bin/rep.exe

//run:
//./bin/rep.exe

#include <zmq.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "common.h"

int main(int argc, char* argv[]) {
    char zmq_version_string[64];
    get_zmq_version_string(zmq_version_string);
    printf("ZMQ version = %s\n", zmq_version_string);

    void* zmq_context = zmq_ctx_new();
    void* reply_socket = zmq_socket(zmq_context, ZMQ_REP);
    if (0 == zmq_bind(reply_socket, "tcp://*:60000")) {
        char buffer[64];
        zmq_recv(reply_socket, buffer, sizeof(buffer), 0);
        printf("%s\n", buffer);
        size_t N = strlen(buffer);
        for (size_t i = 0; i < N; ++i)
            buffer[i] = toupper(buffer[i]);
        zmq_send(reply_socket, buffer, N, 0);
    }
    else {
        printf("%s\n", strerror(errno));
    }

    zmq_close(reply_socket);
    zmq_ctx_term(zmq_context);
    return 0;
}

File: events/visual_studio_summit_2018/src/sample1/Rep/Rep.cs
//Sample provided by Fabio Galuppo  
//May 2018

using System;

class Rep 
{
    static void Main(String[] args) 
    {
        Console.WriteLine("ZMQ version = {0}", Utils.GetZmqVersion());

        var ctx = ZeroMQ.zmq_ctx_new();
        if (ctx != null)
        {
            var replySocket = ZeroMQ.zmq_socket(ctx, ZeroMQ.ZMQ_REP);
            if (replySocket != null)
            {
                if (0 == ZeroMQ.zmq_bind(replySocket, "tcp://*:60000")) 
                {
                    using (var buffer = new PinnedBuffer(new byte[64])) 
                    {
                        if (-1 != ZeroMQ.zmq_recv(replySocket, buffer, buffer.Length, 0)) 
                        {
                            String msg = PinnedBuffer.ASCII(buffer).Replace('\0', ' ').TrimEnd();
                            Console.WriteLine(msg + " received");

                            if (-1 != ZeroMQ.zmq_send(replySocket, buffer, buffer.Length, 0))
                            {
                                Console.WriteLine(msg + " sent back");
                            }
                            else
                            {
                                Console.WriteLine("send failed");
                            }
                        }
                        else
                        {
                            Console.WriteLine("receive failed");
                        }
                    }
                }
                else
                {
                    Console.WriteLine("bind failed");
                }

                ZeroMQ.zmq_close(replySocket);
            }
            else
            {
                Console.WriteLine("socket failed");
            }

            ZeroMQ.zmq_ctx_term(ctx);
        }
        else
        {
            Console.WriteLine("context failed");
        } 
    }
}
